### ISSUE TITLE
Enhance browser history

### ISSUE THREAD

**Issue by alcinnz** [alcinnz/Odysseus#76]
By itself browser history is a near vital feature for a web browser, and one I want to make as useful as possible.

As such upon the basic ability of being able to view which pages I you've visited (at odysseus:history), I want to introduce these other features:

- [x] Highlight changes in the day, month, year, and hour a page was visited. This'll make it quicker to locate a specific time.
- [x] Use history to autocomplete web addresses (shortest and most visited links first). This'll unobtrusively help users remember web addresses in case they get stuck.
- [x] Support searching browser history on that page.
- [x] Upon scrolling odysseus:history render the visible time-range in a consistant location. This'll make it quicker to scroll to a specific time.
- [x] Add favicons
- [x] Show screenshots of recently visited pages upon hovering their link in odysseus:history. This'll help surfers decide whether they've got the right link.
- [x] Illustrate the paths users took through their browsing history. Currently this info is only visible to the (Orca) screen reader, and it'll give more info to locate previously visited pages with.

**Comment by alcinnz** [comment]
Depends on #75, just to have something to build on.

**Comment by alcinnz** [comment]
Having just finished implementing the basics of browser history, I'm quite keen to get on with this.

That said I probably won't fully implement the path rendering (that is I'll probably just write detailed pseudocode) before I move on to implementing topsites. So please join in if you want that.

**Comment by alcinnz** [comment]
I'm now working on designing and researching how to do history search.

Design-wise [history](odysseus:history) is intended to be used when you have some recollection of the page you want to revisit. However I want the search to be a little fuzzy to cater to imperfect memory, and as such implementing this as a Google-style "full text" search would work really well. NOTE: for autocompletion of URIs from history, I'm being more picky so as to limit results and do a better job selecting a default completion.

Then there's the question of "I already have a searchbox visible alongside that page, so do I want to use that?" Previously I've been thinking "yes", but now I've decided the correct answer is "no." Giving the browser history it's own search box clearly distinguishes it's behaviour (and I would want different/additional behaviour on this page), whilst adding something nice for a header.

That said it would be interesting to allow the metasearch engine I plan to implement to include these search results.

Implementation-wise it is easy enough to build a webform submitting a query to the template, which can in turn be passed on to SQLite FTS5. And that would even give me some lovely power user syntax!

**Comment by alcinnz** [comment]
Spent most of my freetime yesterday studying FTS5 and how it works, and I learnt a lot (maybe I'll let myself about it once I've integrated FTS5). However for the purposes of integration: the bad news is that it appears that I'll have to ship it with Odysseus, Debian doesn't offer it; the good news is that it doesn't look hard to do so. I can get it as a (relatively long) C file that'll fit right alongside my build system.

**Comment by alcinnz** [comment]
I've now successfully added FTS5 (for full-text search of browser history) to the install! And I also got any SQL errors that occur to render into the templates, so I'll know if I got something terribly off.

**Comment by alcinnz** [comment]
I've entered panic mode now, as I've realized WebKit can't handle any decent amount of browser history being rendered to a single webpage.

But once I added pagination to address that issue, I found it trivial to add date-range tooltips to them, which nicely addresses the need to make it easier to scroll to a particular date. Hence that aspect's resolved.

**Comment by alcinnz** [comment]
Just tried to embed the full setup inside the odysseus:history template, but (because my code first compiles all the queries (to improve reload efficiency) before any of them are run) it is complaining about the full-text-search table not existing yet.

So I'll need to implement that initialization as a database schema change, which I've been using as an indication to increment the minor version number. So Odysseus 1.2.0 will come out soon with full text search of browser history.

**Comment by alcinnz** [comment]
As of now, I've developed that tool and have successfully used it to simplify some aspects of the history page. Now to see if it helps me deduplicate my database queries!

**Comment by alcinnz** [comment]
Now I'm ready to move on to the next tasks, and I think I'll do something a bit differently from what's suggested on the wiki for communicating your paths through the history.

Previously I was wanting to render lines between pages if you traversed from one to the other and color them based on the tab they were viewed in. But now I'm thinking it'd be cleaner and work better to just color them time labels instead, and add tooltips that communicate from where you navigated to that page. 

This new approach will look cleaner, is easier to implement, and work better with the pagination I had to add.

**Comment by alcinnz** [comment]
Two more to go!

**Comment by alcinnz** [comment]
Adding the favicons was easier then I expected!

Thought I'd have to stuff them into the database and decode it back out. But no, WebKit can handle it. Now I'll have to do that for screenshots, and add a spring cleaning script to avoid cluttering your disk.

**Comment by alcinnz** [comment]
Now I think I've got screenshots saving to the database.

Have to add it to the viewer to see though.

**Comment by alcinnz** [comment]
Added those screenshots into odysseus:history, but they're not showing up in it's HTML source. Investigating now.

**Comment by alcinnz** [comment]
Switched to using GDK Pixbuf and now it works almost magically!

All have to do now is to add a tiny little bit of JavaScript to show the images.

**Comment by alcinnz** [comment]
Got this last item committed, and will definitely release a new version tonight to get this completed vision of browser history out to you (with a couple of fixes).

But I'll leave this issue open until after I get Odysseus to start doing a monthly "spring cleaning," so that your harddrive doesn't fill up with screenshots. I probably want to limit the number of saved screenshots to 10,000.

**Comment by alcinnz** [comment]
Created new issue in regards to the database getting cluttered, so I think I'll close this one.
